File: app/src/main/java/com/ternaryop/photoshelf/parsers/JSONTitleParserConfig.java
package com.ternaryop.photoshelf.parsers;

import java.io.FileInputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

import com.ternaryop.utils.JSONUtils;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Created by dave on 09/01/16.
 * Read from json file
 */
public class JSONTitleParserConfig implements TitleParserConfig {
    private List<TitleParserRegExp> titleCleanerList;
    private Pattern titleParserPattern;
    private List<LocationPrefix> locationPrefixes;
    private HashMap<String, Pattern> cities = new HashMap<>();

    public JSONTitleParserConfig() {
    }

    public JSONTitleParserConfig(String jsonPath) throws Exception {
        readConfig(jsonPath);
    }

    public void readConfig(String jsonPath) throws Exception {
        try (InputStream is = new FileInputStream(jsonPath)) {
            readConfig(is);
        }
    }

    public void readConfig(InputStream is) throws Exception {
        readConfig(JSONUtils.jsonFromInputStream(is));
    }

    public void readConfig(JSONObject jsonObject) throws Exception {
        titleCleanerList = createList(jsonObject, "titleCleaner", "regExprs");
        titleParserPattern = Pattern.compile(jsonObject.getString("titleParserRegExp"), Pattern.CASE_INSENSITIVE);
        readLocationPrefixes(jsonObject);
        readCities(jsonObject);
    }

    private void readCities(JSONObject jsonObject) throws JSONException {
        cities = new HashMap<>();
        final JSONObject jsonCities = jsonObject.getJSONObject("cities");
        final Iterator<String> keys = jsonCities.keys();

        while (keys.hasNext()) {
            final String k = keys.next();
            cities.put(k,  Pattern.compile((String) jsonCities.get(k)));
        }
    }

    private void readLocationPrefixes(JSONObject jsonObject) throws JSONException {
        final JSONArray jsonArray = jsonObject.getJSONArray("locationPrefixes");
        locationPrefixes = new ArrayList<>(jsonArray.length());

        for (int i = 0; i < jsonArray.length(); i++) {
            locationPrefixes.add(new RegExpLocationPrefix(jsonArray.getString(i)));
        }
    }

    @Override
    public List<TitleParserRegExp> getTitleCleanerList() {
        return titleCleanerList;
    }

    @Override
    public String applyList(List<TitleParserRegExp> titleParserRegExpList, String input) {
        return TitleParserRegExp.applyList(titleParserRegExpList, input);
    }

    @Override
    public Pattern getTitleParserPattern() {
        return titleParserPattern;
    }

    @Override
    public List<LocationPrefix> getLocationPrefixes() {
        return locationPrefixes;
    }

    @Override
    public Map<String, Pattern> getCities() {
        return cities;
    }

    public static ArrayList<TitleParserRegExp> createList(JSONObject jsonAssets, String rootName, String replacers) throws Exception {
        ArrayList<TitleParserRegExp> list = new ArrayList<>();
        JSONArray array = jsonAssets.getJSONObject(rootName).getJSONArray(replacers);
        for (int i = 0; i < array.length(); i++) {
            JSONArray reArray = array.getJSONArray(i);
            list.add(new TitleParserRegExp(reArray.getString(0), reArray.getString(1)));
        }
        return list;
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/parsers/JSONTitleParserConfig.kt
package com.ternaryop.photoshelf.parsers

import com.ternaryop.utils.json.readJson
import org.json.JSONException
import org.json.JSONObject
import java.io.FileInputStream
import java.util.regex.Pattern

/**
 * Created by dave on 09/01/16.
 * Read from json file
 */
class JSONTitleParserConfig : TitleParserConfig {
    override lateinit var titleCleanerList: List<TitleParserConfig.TitleParserRegExp>
        private set
    override lateinit var titleParserPattern: Pattern
        private set
    override lateinit var locationPrefixes: List<LocationPrefix>
        private set
    override lateinit var cities: Map<String, Pattern>
        private set

    constructor()

    @Throws(Exception::class) constructor(jsonPath: String) {
        readConfig(jsonPath)
    }

    @Throws(Exception::class)
    private fun readConfig(jsonPath: String) {
        FileInputStream(jsonPath).use { stream -> readConfig(stream.readJson()) }
    }

    @Throws(Exception::class)
    fun readConfig(jsonObject: JSONObject) {
        titleCleanerList = createList(jsonObject, "titleCleaner", "regExprs")
        titleParserPattern = Pattern.compile(jsonObject.getString("titleParserRegExp"), Pattern.CASE_INSENSITIVE)
        locationPrefixes = readLocationPrefixes(jsonObject)
        cities = readCities(jsonObject)
    }

    @Throws(JSONException::class)
    private fun readCities(jsonObject: JSONObject): Map<String, Pattern> {
        val map: HashMap<String, Pattern> = HashMap()
        val jsonCities = jsonObject.getJSONObject("cities")
        val keys = jsonCities.keys()

        while (keys.hasNext()) {
            val k = keys.next() as String
            map[k] = Pattern.compile(jsonCities.getString(k))
        }
        return map
    }

    @Throws(JSONException::class)
    private fun readLocationPrefixes(jsonObject: JSONObject): List<LocationPrefix> {
        val jsonArray = jsonObject.getJSONArray("locationPrefixes")

        return (0 until jsonArray.length()).map { RegExpLocationPrefix(jsonArray.getString(it)) }
    }

    override fun applyList(titleParserRegExpList: List<TitleParserConfig.TitleParserRegExp>, input: String): String {
        return TitleParserConfig.TitleParserRegExp.applyList(titleParserRegExpList, input)
    }

    companion object {

        @Throws(Exception::class)
        fun createList(jsonAssets: JSONObject, rootName: String, replacers: String):
            MutableList<TitleParserConfig.TitleParserRegExp> {
            val list = mutableListOf<TitleParserConfig.TitleParserRegExp>()
            val array = jsonAssets.getJSONObject(rootName).getJSONArray(replacers)
            for (i in 0 until array.length()) {
                val reArray = array.getJSONArray(i)
                list.add(TitleParserConfig.TitleParserRegExp(reArray.getString(0), reArray.getString(1)))
            }
            return list
        }
    }
}
